database/pebble/db.go

```diff
@@ -0,0 +1,341 @@
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	switch {
+	case db.closed:
+		return database.ErrClosed
+	case comparer.Compare(start, limit) >= 0:
+		// pebble's Compact will no-op & error if start >= limit
+		// according to pebble's comparer.
+		return nil
+	case limit != nil:
+		return updateError(db.pebbleDB.Compact(start, limit, true /* parallelize */))
+	}
+
+	// The database.Database spec treats a nil [limit] as a key after all keys
+	// but pebble treats a nil [limit] as a key before all keys.
+	// Use the greatest key in the database as the [limit] to get the desired behavior.
+	it := db.pebbleDB.NewIter(&pebble.IterOptions{})
+	if it.Last() {
+		if lastkey := it.Key(); lastkey != nil {
+			return updateError(db.pebbleDB.Compact(start, lastkey, true /* parallelize */))
+		}
+	}
+
+	// Either this database is empty or the only key in it is nil.
+	return nil
+}
+
+func (db *Database) NewIterator() database.Iterator {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	if db.closed {
+		return &iter{
+			db:     db,
+			closed: true,
+			err:    database.ErrClosed,
+		}
+	}
+
+	iter := &iter{
+		db:   db,
+		iter: db.pebbleDB.NewIter(&pebble.IterOptions{}),
+	}
+	db.openIterators.Add(iter)
+	return iter
+}
+
+func (db *Database) NewIteratorWithStart(start []byte) database.Iterator {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	if db.closed {
+		return &iter{
+			db:     db,
+			closed: true,
+			err:    database.ErrClosed,
+		}
+	}
+
+	iter := &iter{
+		db: db,
+		iter: db.pebbleDB.NewIter(&pebble.IterOptions{
+			LowerBound: start,
+		}),
+	}
+	db.openIterators.Add(iter)
+	return iter
+}
+
+func (db *Database) NewIteratorWithPrefix(prefix []byte) database.Iterator {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	if db.closed {
+		return &iter{
+			db:     db,
+			closed: true,
+			err:    database.ErrClosed,
+		}
+	}
+
+	iter := &iter{
+		db:   db,
+		iter: db.pebbleDB.NewIter(prefixBounds(prefix)),
+	}
+	db.openIterators.Add(iter)
+	return iter
+}
+
+func (db *Database) NewIteratorWithStartAndPrefix(start, prefix []byte) database.Iterator {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	if db.closed {
+		return &iter{
+			db:     db,
+			closed: true,
+			err:    database.ErrClosed,
+		}
+	}
+
+	iterRange := prefixBounds(prefix)
+	if bytes.Compare(start, prefix) == 1 {
+		iterRange.LowerBound = start
+	}
+	iter := &iter{
+		db:   db,
+		iter: db.pebbleDB.NewIter(iterRange),
+	}
+	db.openIterators.Add(iter)
+	return iter
+}
+
+// Converts a pebble-specific error to to its
+// Avalanche equivalent, if applicable.
+func updateError(err error) error {
+	switch err {
+	case pebble.ErrClosed:
+		return database.ErrClosed
+	case pebble.ErrNotFound:
+		return database.ErrNotFound
+	default:
+		return err
+	}
+}
+
+// Returns a key range that covers all keys with the
+// given [prefix].
+// Assumes the Database uses bytes.Compare for key
+// comparison and not a custom comparer.
+func prefixBounds(prefix []byte) *pebble.IterOptions {
```

Comment:
Where did this come from?

Reply:
I'm not sure what you mean. This is how we generate iteration bounds when we use a prefix.